### PR TITLE
GQA Flash Attention with Attention Mask

### PR DESCRIPTION
**Pull request by aciddelgado**
### Description
GQA now only works with Flash Attention with Attention Mask input, allowing for batched input. Note: This PR Disables Memory Efficient Attention, only allowing Flash Attention kernel to be used.

### Motivation and Context
Allows GQA to work with batched input.

### CODE REVIEW

**Comment by tianleiwu** [comment]
Please resolve conflicts

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 1]
```diff
@@ -1037,9 +1037,8 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
                "T",
                OpSchema::Optional)
         .Input(5,
-               "past_sequence_length",
-               "When buffered past_key and past_value is used (present_key uses same tensor as past_key), required"
-               "to specify past_sequence_length (could be 0). Otherwise, past_sequence_length inferred from past_key.",
+               "attention_mask",
+               "2d Tensor of shape (batch_size, past_sequence_length + sequence_length). Must be a right padding mask.",
```
Comment (tianleiwu) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc:1041]: Could we support both left side and right side padding?

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 2]
```diff
@@ -1000,14 +1000,14 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
         .SetDoc(GroupQueryAttention_ver1_doc)
         .Attr("num_heads", "Number of attention heads for q", AttributeProto::INT)
         .Attr("kv_num_heads", "Number of attention heads for k and v", AttributeProto::INT)
-        .Attr("unidirectional",
```
Comment (tianleiwu) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc:1003]: why remove unidirectional attribute?

File: docs/ContribOperators.md [hunk 3]
```diff
@@ -2421,14 +2421,14 @@ This version of the operator has been available since version 1 of the 'com.micr
 <dd>When buffered past_key and past_value is used (present_key uses same tensor as past_key), requiredto specify past_sequence_length (could be 0). Otherwise, past_sequence_length inferred from past_key.</dd>
```
Comment (tianleiwu) [inline docs/ContribOperators.md:2421]: Need a space between required and to

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc (hunk 2 above)
Comment (aciddelgado) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc:1003]: It seems the use case for GQA for now always includes causal

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc (hunk 1 above)
Comment (aciddelgado) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc:1041]: Flash Attention does not support left padding unfortunately

Comment (tianleiwu) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc:1041]: It's fine to throw Not Implemented exception in CUDA. For the interface, we can describe that it supports left padding (which is needed by DML).

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc (hunk 2 above)
Comment (tianleiwu) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc:1003]: It's fine to leave it. At least it is very clear that it is causal, and consistent with other operators.
You can throw Not Implemented if non causal is not supported in kernel.

File: onnxruntime/contrib_ops/cpu/transformers/logits_processor.h [hunk 4]
```diff
@@ -266,9 +266,9 @@ class TimestampLogitsProcessor : public ILogitsProcessor<T> {
       }
     }
 
-#ifdef DEBUG_GENERATION
-    DumpScores("TimestampLogitsProcessor", next_token_scores);
-#endif
+    // #ifdef DEBUG_GENERATION
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/logits_processor.h:269]: revert this file.

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 5]
```diff
@@ -1059,7 +1058,7 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
                 "kv_sequence_length.",
                 "T")
         .TypeConstraint("T", {"tensor(float16)"}, "Constrain input and output to float tensors.")
-        .TypeConstraint("M", {"tensor(int32)", "tensor(int64)"}, "Constrain past sequence length to int tensor.")
+        .TypeConstraint("M", {"tensor(int64)"}, "Constrain past sequence length to int tensor.")
```
Comment (tianleiwu) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc:1061]: update the description since there is no past sequence length in input now. Check also the doc for operator.

File: cmake/external/onnx
Comment (yufenglee) [inline cmake/external/onnx:1]: please revert the change of onnx

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc (hunk 1 above)
Comment (yufenglee) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc:1041]: we can do this later, but not for this bug fix.

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 6]
```diff
@@ -1000,14 +1000,14 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
...
-              AttributeProto::INT,
-              static_cast<int64_t>(1))
         .Attr("is_past_bsnh",
               "Whether past kv uses BSNH, otherwise BNSH. Default value is 1 (BSNH).",
               AttributeProto::INT,
               static_cast<int64_t>(1))
+        .Attr("kv_share_buffer",
+              "Whether past kv shares tensor with present kv.",
+              AttributeProto::INT,
+              static_cast<int64_t>(1))
```
Comment (yufenglee) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc:1010]: why do we need this? User needs to generate different models. we can get this just from if past_kv and present_kv have same address.

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.cu [hunk 7]
```diff
@@ -417,6 +440,117 @@
...
+  // Specialize BlockReduce type for our thread block
+  typedef cub::BlockReduce<int, BLOCK_THREADS> BlockReduceT;
+  __shared__ typename BlockReduceT::TempStorage temp_storage;
+
+  const int b = blockIdx.x;
+	int mask_offset = b * mask_seqlen;
+  int sum = 0;
+	for(int i = threadIdx.x; i < mask_seqlen; i += blockDim.x) {
+		sum += attention_mask[mask_offset + i];
+	}
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.cu:458]: code style

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc [hunk 8]
```diff
@@ -120,23 +120,19 @@
...
-  auto seqlens_k_buffer = GetScratchBuffer<void>(seqlens_k_bytes, context->GetComputeStream());
 #else
   constexpr bool use_flash_attention = false;
   auto softmax_lse_buffer = GetScratchBuffer<void>(0, context->GetComputeStream());        // nullptr
   auto softmax_lse_accum_buffer = GetScratchBuffer<void>(0, context->GetComputeStream());  // nullptr
   auto out_accum_buffer = GetScratchBuffer<void>(0, context->GetComputeStream());          // nullptr
-  auto seqlens_k_buffer = GetScratchBuffer<void>(0, context->GetComputeStream());          // nullptr
 #endif
 
+  ORT_ENFORCE(use_flash_attention);
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc:134]: add a message like "GroupQueryAttention is only supported on ampere gpu and above"

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 9]
```diff
@@ -1037,9 +1037,8 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
                "T",
                OpSchema::Optional)
         .Input(5,
-               "past_sequence_length",
-               "When buffered past_key and past_value is used (present_key uses same tensor as past_key), required"
-               "to specify past_sequence_length (could be 0). Otherwise, past_sequence_length inferred from past_key.",
+               "attention_mask",
+               "2d Tensor of shape (batch_size, past_sequence_length + sequence_length). Must be a right padding mask.",
                "M",
                OpSchema::Optional)
```
Comment (yufenglee) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc:1043]: why not make attention_mask a required parameter?

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc [hunk 10]
```diff
@@ -198,8 +216,24 @@
...
+    if (seqstart_q_buffer != nullptr) {
+      data.seqstart_q = reinterpret_cast<int32_t*>(seqstart_q_buffer.get());
+    }
+  }
+  if (k_buffer != nullptr) {
+    data.k = reinterpret_cast<CudaT*>(k_buffer.get());
+    data.v = reinterpret_cast<CudaT*>(v_buffer.get());
+  }
+  if (fmha_buffer != nullptr) {
+    data.fmha_buffer = reinterpret_cast<CudaT*>(fmha_buffer.get());
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc:236]: duplicated with code below.

File: onnxruntime/contrib_ops/cpu/bert/attention_common.h [hunk 11]
```diff
@@ -91,13 +91,17 @@ struct GroupQueryAttentionParameters {
...
   int max_sequence_length;      // allocated length of past_key and past_value
+  int mask_sequence_length;     // sequence length of mask (should be max past + max new sequence lengths)
   int hidden_size;
   int num_heads;
   int head_size;
   int kv_hidden_size;
   int kv_num_heads;
-  int num_splits;          // number of splits for splitkv
+  int num_splits;  // number of splits for splitkv
+  bool has_mask;
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/bert/attention_common.h:101]: we don't need it anymore as it is always true. it will make the code cleaner.

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc [hunk 12]
```diff
@@ -151,22 +148,44 @@
...
   auto k_buffer = GetScratchBuffer<void>(0, context->GetComputeStream());
   auto v_buffer = GetScratchBuffer<void>(0, context->GetComputeStream());
   auto fmha_buffer = GetScratchBuffer<void>(0, context->GetComputeStream());
+  auto seqstart_k_buffer = GetScratchBuffer<void>(0, context->GetComputeStream());
+  auto seqstart_q_buffer = GetScratchBuffer<void>(0, context->GetComputeStream());
 #endif
 
+  // seqlens_k buffer
+  size_t seqlens_k_bytes = 0;
+  if (past_key != nullptr || parameters.has_mask) {
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc:184]: >parameters.has_mask [](http://example.com/codeflow?start=29&length=19)

as has_mask is always true, you can clean up the code accordingly

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc [hunk 13]
```diff
@@ -208,6 +236,16 @@
...
+  if (k_buffer != nullptr) {
+    data.k = reinterpret_cast<CudaT*>(k_buffer.get());
+    data.v = reinterpret_cast<CudaT*>(v_buffer.get());
+  }
+
+  if (data.past_key == data.present_key) {
+    parameters.kv_share_buffer = true;
+  } else {
+    parameters.kv_share_buffer = false;
+  }
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc:248]: nit: parameters.kv_share_buffer = data.past_key == data.present_key;

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention_helper.h [hunk 14]
```diff
@@ -19,7 +19,7 @@
                    void* parameters,
                    int num_heads,
                    int kv_num_heads,
-                   const Tensor* past_seq_len,
+                   const Tensor* attention_mask,
                    bool is_past_bsnh,
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention_helper.h:23]: >                  bool is_past_bsnh, [](http://example.com/codeflow?start=1&length=36)

as we only support bnsh, you can clean up the code

Comment (aciddelgado) [reply onnxruntime/contrib_ops/cuda/bert/group_query_attention_helper.h:23]: i thought i'd leave it since everything is in place to support bsnh

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention_helper.h [hunk 15]
```diff
@@ -47,10 +47,10 @@ Status CheckInputs(const Tensor* query,
   int q_hidden_size = static_cast<int>(query_dims[2]);
   int head_size = static_cast<int>(q_hidden_size) / num_heads;
 
-  int kv_sequence_length = static_cast<int>(key_dims[1]);
   int kv_hidden_size = static_cast<int>(key_dims[2]);
 
-  int max_sequence_length = 0;
+  int32_t past_sequence_length = 0;
+  int cache_sequence_length = 0;
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention_helper.h:53]: duplicated. only need to keep one

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc [hunk 16]
```diff
@@ -108,72 +107,85 @@ Status GroupQueryAttention<T>::ComputeInternal(OpKernelContext* context) const {
...
   size_t out_accum_bytes = 0;
-  size_t seqlens_k_bytes = 0;
   if (use_flash_attention) {
     // softmax buffer
     softmax_lse_bytes = onnxruntime::flash::get_softmax_lse_size(parameters.sequence_length, parameters.batch_size, parameters.num_heads);
     // split kv buffer
     using namespace std;
     auto [num_splits, slse_accum_bytes, o_accum_bytes] = onnxruntime::flash::get_num_splits_and_buffer_sizes(
-        parameters.batch_size, parameters.sequence_length, parameters.kv_sequence_length, parameters.num_heads,
+        parameters.batch_size, parameters.sequence_length, parameters.sequence_length, parameters.num_heads,
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc:116]: >sequence_length [](http://example.com/codeflow?start=70&length=15)

is it mask_sequence_length?

File: onnxruntime/contrib_ops/cpu/transformers/logits_processor.h [hunk 17]
```diff
@@ -266,9 +266,9 @@ class TimestampLogitsProcessor : public ILogitsProcessor<T> {
       }
     }
 
-#ifdef DEBUG_GENERATION
-    DumpScores("TimestampLogitsProcessor", next_token_scores);
-#endif
+// #ifdef DEBUG_GENERATION
+//     DumpScores("TimestampLogitsProcessor", next_token_scores);
+// #endif
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/transformers/logits_processor.h:271]: it is still here. you may need to revert the change.

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc (hunk 16 above)
Comment (aciddelgado) [reply onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc:116]: i believe it is the normal one

File: onnxruntime/contrib_ops/cpu/bert/attention_common.h [hunk 18]
```diff
@@ -86,18 +86,25 @@ struct PackedAttentionParameters {
 // Parameters deduced from node attributes and inputs/outputs.
 struct GroupQueryAttentionParameters {
   int batch_size;
-  int sequence_length;
-  int past_sequence_length;     // actual sequence length of past_key and past_value
-  int kv_sequence_length;       // sequence length of key and value (or new_k and new_v when past is present)
-  int present_sequence_length;  // past_sequence_length + kv_sequence_length
-  int max_sequence_length;      // allocated length of past_key and past_value
+  // int sequence_length;
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/bert/attention_common.h:89]: remove those commented out code after verification

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention_helper.h [hunk 19]
```diff
@@ -129,8 +126,12 @@ Status CheckInputs(const Tensor* query,
     }
   } else if (past_key != nullptr || past_value != nullptr) {
     return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
-                           "Input 'past_key' and 'past_value' shall be both present or both absent");
+                           "Input 'past_key' and 'past_value' shall be both present or both absent.");
   }
+  // } else if (kv_share_buffer && past_key == nullptr && past_value == nullptr) {
+  //   return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+  //                          "Input 'past_key' and 'past_value' shall be present when kv_share_buffer is on.");
+  // }
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention_helper.h:134]: nit: remove

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc (hunk 16 above)
Comment (yufenglee) [reply onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc:116]: It shall be the present sequece length. The idea is to increase the number of block by dividing on K and V:
[https://github.com/Dao-AILab/flash-attention/blob/83aef842beec1037eb8c1d9c3ef3ed8aae80b091/csrc/flash_attn/src/flash_fwd_launch_template.h#L76](https://github.com/Dao-AILab/flash-attention/blob/83aef842beec1037eb8c1d9c3ef3ed8aae80b091/csrc/flash_attn/src/flash_fwd_launch_template.h#L76)

File: cmake/deps.txt [hunk 20]
```diff
@@ -16,7 +16,7 @@ abseil_cpp;https://github.com/abseil/abseil-cpp/archive/refs/tags/20230802.0.zip
 cxxopts;https://github.com/jarro2783/cxxopts/archive/3c73d91c0b04e2b59462f0a741be8c07024c1bc0.zip;6c6ca7f8480b26c8d00476e0e24b7184717fe4f0
 date;https://github.com/HowardHinnant/date/archive/refs/tags/v3.0.1.zip;2dac0c81dc54ebdd8f8d073a75c053b04b56e159
 dlpack;https://github.com/dmlc/dlpack/archive/refs/tags/v0.6.zip;4d565dd2e5b31321e5549591d78aa7f377173445
-eigen;https://gitlab.com/libeigen/eigen/-/archive/3.4/eigen-3.4.zip;ee201b07085203ea7bd8eb97cbcb31b07cfa3efb
+eigen;https://gitlab.com/libeigen/eigen/-/archive/3.4.0/eigen-3.4.0.zip;ef24286b7ece8737c99fa831b02941843546c081
```
Comment (tianleiwu) [inline cmake/deps.txt:19]: Please merge main (I think it is resolved in main branch).

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.cu [hunk 21]
```diff
@@ -608,6 +753,10 @@ Status QkvToContext(
   }
 #endif
 
+  if (parameters.left_padding && parameters.is_prompt) {
+    ORT_RETURN_IF_ERROR(LaunchLeftPadLast(parameters, data, stream, device_prop.maxThreadsPerBlock));
+  }
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.cu:758]: It won't be called

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.cu [hunk 22]
```diff
@@ -575,10 +719,11 @@ Status EfficientAttention(
                     ? data.fmha_buffer
                     : nullptr;
   p.stream = stream;
+  p.has_custom_right_padding = true;
   run_memory_efficient_attention(p);
 
-  DUMP_TENSOR_INIT();
-  DUMP_TENSOR("efficient attention output", data.output, batch_size, sequence_length, num_heads, head_size);
+  // DUMP_TENSOR_INIT();
+  // DUMP_TENSOR("efficient attention output", data.output, batch_size, sequence_length, num_heads, head_size);
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.cu:727]: uncomment those 2 lines

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.cu [hunk 23]
```diff
@@ -262,44 +283,47 @@ Status LaunchConcatKVInPlace(contrib::GroupQueryAttentionParameters& parameters,
...
-  const int kv_sequence_length = parameters.kv_sequence_length;
-  const int present_sequence_length = parameters.present_sequence_length;
-  const int past_sequence_length = parameters.past_sequence_length;
+  const int kv_sequence_length = parameters.sequence_length;
+  const int present_sequence_length = parameters.seqlen_present_kv_cache;
   const int kv_num_heads = parameters.kv_num_heads;
   const int head_size = parameters.head_size;
+
+  // Indicates past sequence_length of each sequence
+  const int* seqlens_k = parameters.is_prompt ? nullptr : reinterpret_cast<const int*>(data.seqlens_k_buff);
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.cu:292]: >reinterpret_cast<const int*>(data.seqlens_k_buff); [](http://example.com/codeflow?start=58&length=50)

just use data.seqlens_k

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.cu [hunk 24]
```diff
@@ -443,42 +543,78 @@
 
   bool is_causal = parameters.is_unidirectional;
 
-  if (data.past_key != nullptr && data.past_key == data.present_key) {
+  // Note: seqlens_k is past sequence length for flash
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.cu:546]: we have 2 seqlens_k here: one is data.seqlens_k, the other is data_selens_k_buff. I think we can make it more clean. 
data.seqlens_k to store the total_sequence - 1 and data.seqlens_k_buff to store the total_sequence (also change its name to data.seqlens_k_total). It will make the handling below easier.

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.h [hunk 25]
```diff
@@ -20,15 +20,18 @@ struct GroupQueryAttentionData {
...
   T* softmax_lse_accum = nullptr;
   T* out_accum = nullptr;
-  int* seqlens_k = nullptr;
+  int* seqlens_k_total = nullptr;
   // Memory Efficient buffers
   T* fmha_buffer = nullptr;
   T* k = nullptr;
   T* v = nullptr;
+  int32_t* seqstart_q = nullptr;
+  int32_t* seqstart_k = nullptr;
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.h:34]: nit: we don't need it anymore. We can clean it up later.

**Review by yufenglee — APPROVED** [review]
:shipit:
